### PR TITLE
planner: fix subquery update with generated column

### LINKED ISSUE #23947: update join paniced with generated column tables
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

create 3 empty tables

```
CREATE TABLE `topics` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `title` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `last_posted_at` datetime DEFAULT NULL,
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  `views` int(11) NOT NULL DEFAULT '0',
  `posts_count` int(11) NOT NULL DEFAULT '0',
  `user_id` int(11) DEFAULT NULL,
  `last_post_user_id` int(11) NOT NULL,
  `reply_count` int(11) NOT NULL DEFAULT '0',
  `featured_user1_id` int(11) DEFAULT NULL,
  `featured_user2_id` int(11) DEFAULT NULL,
  `featured_user3_id` int(11) DEFAULT NULL,
  `avg_time` int(11) DEFAULT NULL,
  `deleted_at` datetime DEFAULT NULL,
  `highest_post_number` int(11) NOT NULL DEFAULT '0',
  `image_url` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `like_count` int(11) NOT NULL DEFAULT '0',
  `incoming_link_count` int(11) NOT NULL DEFAULT '0',
  `category_id` int(11) DEFAULT NULL,
  `visible` tinyint(1) NOT NULL DEFAULT '1',
  `moderator_posts_count` int(11) NOT NULL DEFAULT '0',
  `closed` tinyint(1) NOT NULL DEFAULT '0',
  `archived` tinyint(1) NOT NULL DEFAULT '0',
  `bumped_at` datetime NOT NULL,
  `has_summary` tinyint(1) NOT NULL DEFAULT '0',
  `archetype` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT 'regular',
  `featured_user4_id` int(11) DEFAULT NULL,
  `notify_moderators_count` int(11) NOT NULL DEFAULT '0',
  `spam_count` int(11) NOT NULL DEFAULT '0',
  `pinned_at` datetime DEFAULT NULL,
  `score` float DEFAULT NULL,
  `percent_rank` float NOT NULL DEFAULT '1',
  `subtype` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `slug` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `deleted_by_id` int(11) DEFAULT NULL,
  `participant_count` int(11) DEFAULT '1',
  `word_count` int(11) DEFAULT NULL,
  `excerpt` varchar(1000) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `pinned_globally` tinyint(1) NOT NULL DEFAULT '0',
  `pinned_until` datetime DEFAULT NULL,
  `fancy_title` varchar(400) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `highest_staff_post_number` int(11) NOT NULL DEFAULT '0',
  `featured_link` varchar(255) COLLATE utf8mb4_unicode_ci DEFAULT NULL,
  `reviewable_score` float NOT NULL DEFAULT '0',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  KEY `index_topics_on_bumped_at` (`bumped_at`),
  KEY `index_topics_on_created_at_and_visible` (`created_at`,`visible`),
  KEY `index_topics_on_id_and_deleted_at` (`id`,`deleted_at`),
  KEY `index_topics_on_pinned_at` (`pinned_at`),
  KEY `index_topics_on_pinned_globally` (`pinned_globally`),
  KEY `index_topics_on_updated_at_public` (`updated_at`,`visible`,`highest_staff_post_number`,`highest_post_number`,`category_id`,`created_at`,`id`),
  KEY `idx_topics_user_id_deleted_at` (`user_id`),
  KEY `idx_topics_front_page_2` (`deleted_at`,`visible`,`archetype`,`category_id`,`id`),
  KEY `idxtopicslug2` (`slug`)
);


CREATE TABLE `topic_tags` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `topic_id` int(11) NOT NULL,
  `tag_id` int(11) NOT NULL,
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `index_topic_tags_on_topic_id_and_tag_id` (`topic_id`,`tag_id`)
);

CREATE TABLE `tags` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `name` varchar(255) COLLATE utf8mb4_unicode_ci NOT NULL,
  `topic_count` int(11) NOT NULL DEFAULT '0',
  `created_at` datetime NOT NULL,
  `updated_at` datetime NOT NULL,
  `pm_topic_count` int(11) NOT NULL DEFAULT '0',
  `lower_tag_name` varchar(255) COLLATE utf8mb4_unicode_ci GENERATED ALWAYS AS (lower(`name`)) STORED,
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  UNIQUE KEY `index_tags_on_name` (`name`),
  UNIQUE KEY `index_tags_on_lower_name` (`lower_tag_name`)
);
```

update with

```
UPDATE tags t INNER JOIN (SELECT COUNT(topics.id) AS topic_count, tags.id AS tag_id FROM tags LEFT JOIN topic_tags ON tags.id = topic_tags.tag_id LEFT JOIN topics ON topics.id = topic_tags.topic_id AND topics.deleted_at IS NULL AND topics.archetype != 'private_message' GROUP BY tags.id) x ON x.tag_id = t.id AND x.topic_count <> t.topic_count SET t.topic_count = x.topic_count;
```

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): runtime error: index out of range [-1]
```

```
goroutine 139229 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0x10733d9f8, 0x1401154b980, 0x140117f7500)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:734 +0x90
panic(0x1070e87a0, 0x1400f760600)
	/usr/local/go/src/runtime/panic.go:971 +0x424
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildUpdateLists(0x1400ffb14a0, 0x10733d9f8, 0x140100193b0, 0x140101185d0, 0x2, 0x2, 0x140124d4020, 0x1, 0x1, 0x1073689d0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:4477 +0x21c8
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildUpdate(0x1400ffb14a0, 0x10733d9f8, 0x140100193b0, 0x1400fe30e80, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:4303 +0x1154
github.com/pingcap/tidb/planner/core.(*PlanBuilder).Build(0x1400ffb14a0, 0x10733d9f8, 0x140100193b0, 0x10734e3f8, 0x1400fe30e80, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/core/planbuilder.go:630 +0xc88
github.com/pingcap/tidb/planner.optimize(0x10733d9f8, 0x140100193b0, 0x1073696d8, 0x140117f7600, 0x10734e3f8, 0x1400fe30e80, 0x107364608, 0x140122ee230, 0x0, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/optimize.go:234 +0x1fc
github.com/pingcap/tidb/planner.Optimize(0x10733d9f8, 0x140100193b0, 0x1073696d8, 0x140117f7600, 0x10734e3f8, 0x1400fe30e80, 0x107364608, 0x140122ee230, 0x0, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/planner/optimize.go:124 +0x70c
github.com/pingcap/tidb/executor.(*Compiler).Compile(0x140101caba8, 0x10733d9f8, 0x140100193b0, 0x107352a48, 0x1400fe30e80, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/executor/compiler.go:62 +0x444
github.com/pingcap/tidb/session.(*session).ExecuteStmt(0x140117f7600, 0x10733d9f8, 0x14010019230, 0x107352a48, 0x1400fe30e80, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/session/session.go:1403 +0x520
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt(0x1401154bdd0, 0x10733d9f8, 0x14010019200, 0x107352a48, 0x1400fe30e80, 0x0, 0x0, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/driver_tidb.go:218 +0x6c
github.com/pingcap/tidb/server.(*clientConn).handleStmt(0x140117f7500, 0x10733d9f8, 0x14010019200, 0x107352a48, 0x1400fe30e80, 0x108b42af8, 0x0, 0x0, 0x1, 0x0, ...)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:1623 +0x118
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0x140117f7500, 0x10733d950, 0x1400fdb3540, 0x1401ac9e481, 0x179, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:1496 +0x800
github.com/pingcap/tidb/server.(*clientConn).dispatch(0x140117f7500, 0x10733d950, 0x1400fdb3540, 0x1401ac9e481, 0x179, 0x179, 0x0, 0x0)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:1030 +0xc04
github.com/pingcap/tidb/server.(*clientConn).Run(0x140117f7500, 0x10733d9f8, 0x1401154b980)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/conn.go:795 +0x7bc
github.com/pingcap/tidb/server.(*Server).onConn(0x1400fa188f0, 0x140117f7500)
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/server.go:477 +0x5ec
created by github.com/pingcap/tidb/server.(*Server).Run
	/Users/robi/Code/go/src/github.com/pingcap/tidb/server/server.go:380 +0x610
```

### 4. What is your TiDB version? (Required)

5.0.0

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: close #23947 

Problem Summary:

generate column will generate assign to non-updatable subquery part

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed, How it Works:

don't generate assign to not updatable subquery

### Related changes

- Need to cherry-pick to the release branch 5.0

### Check List 

Tests 

- Unit test

Side effects

- n/a

### Release note 

- Fix panic when subquery update stmt the table with generated columns

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/23955)

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -4431,8 +4423,17 @@ func (b *PlanBuilder) buildUpdateLists(
 	// If columns in set list contains generated columns, raise error.
 	// And, fill virtualAssignments here; that's for generated columns.
 	virtualAssignments := make([]*ast.Assignment, 0)
-
 	for _, tn := range tableList {
+		updatable := true
+		for _, nTbl := range notUpdatableTbl {
```
Comment (wjhuang2016) [inline planner/core/logical_plan_builder.go:4428]: Add some comments here?

**Comment by cfzjywxk** [comment]
LGTM

**Comment by cfzjywxk** [comment]
This issue does not exist on release-4.0?

**Comment by lysu** [comment]
> This issue does not exist on release-4.0?

yes, it's introduced by https://github.com/pingcap/tidb/pull/23404 and only affect on 5.0

**Review by cfzjywxk — APPROVED** [review]
/lgtm

**Pull request merged**